vm-setup/roles/v1aX_integration_test/tasks/node_reuse.yml

```diff
@@ -9,117 +9,132 @@
-    register: provisioning_machine
-
   - name: Check if only single machine is in Deleting state and no other new machine is in Provisioning state while upgrade triggered to test the KCP scale-in feature.
-    shell: |
-        kubectl get machines -n "{{ NAMESPACE }}" -o json | jq -r '[ .items[] | select (.status.phase == "Deleting" or .status.phase == "deleting") | .metadata.name ] | length'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: deleting_machines
+    k8s_info:
+      api_version: cluster.x-k8s.io/v1alpha3
+      kind: machines
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: provisioning_machine
     retries: 200
     delay: 2
-    until: deleting_machines.stdout == NUMBER_OF_BMH_MD
-    failed_when: provisioning_machine.stdout != "0"
+    until: provisioning_machine.resources | filter_phase("deleting") | length == (NUMBER_OF_BMH_MD | int)
+    failed_when: provisioning_machine.resources | filter_phase("Provisioning") | length  != 0
 
   - name: Wait until "{{ NUMBER_OF_BMH_MD }}" BMH is in deprovisioning state.
-    shell: |
-        kubectl get bmh -n "{{ NAMESPACE }}" -o json | jq -r '[ .items[]
-        | select (.status.provisioning.state == "deprovisioning")
-        | .metadata.name ] | length'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: bmh_in_deprovisioning
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: deprovisioning_bmh
     retries: 150
     delay: 10
-    until: bmh_in_deprovisioning.stdout == NUMBER_OF_BMH_MD
-
-  - name: Get the name of the deprovisioning BMH.
-    shell: |
-        kubectl get bmh -n "{{ NAMESPACE }}" -o json | jq -r '[ .items[]
-        | select (.status.provisioning.state == "deprovisioning")
-        | .metadata.name ] | add'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: deprovisioning_bmh_name
+    until: deprovisioning_bmh.resources | filter_provisioning("deprovisioning") | length == (NUMBER_OF_BMH_MD | int)
 
   - name: Wait until above deprovisioning BMH is in Ready state again.
-    shell: |
-        kubectl get bmh -n "{{ NAMESPACE }}" | grep -w ready | awk '{print $1}'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
     register: ready_bmh
     retries: 200
     delay: 2
-    until: deprovisioning_bmh_name.stdout in ready_bmh.stdout_lines
+    until: (deprovisioning_bmh.resources | filter_provisioning("deprovisioning") | get_names) |
+           intersect(ready_bmh.resources | filter_provisioning("ready") | get_names) |
+           length ==  (NUMBER_OF_BMH_MD | int)
```

Comment:
Here if I well followed we are trying to make sure that the set of BMHs `n(BMHs) = NUMBER_OF_BMH_MD ` that was in `deprovisioning` state is now in `ready` state knowing that I tried to intersect the two sets (deprovisioning ⋂ ready ) and ensure that the intersect equals the deprovisioning length (that's equivalent to `deprovisioning_names difference(ready_name) == 0 )`.

I will see if it can be simplified